### PR TITLE
Protect against invalid file uploads

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes https://github.com/metabase/metabase-private/issues/227

### Description

This PR adds file type validation to the CSV upload API endpoints.

### CODE REVIEW

File: src/metabase/upload.clj [hunk 1]
```diff
@@ -215,12 +216,25 @@
       (throw (ex-info (tru "Unable to recognise file separator")
                       {:status-code 422}))))
 
+(def ^:private allowed-extensions #{nil "csv" "tsv" "txt"})
+
+(def ^:private allowed-mime-types #{"text/csv" "text/tab-separated-values" "text/plain"})
+
+(def ^:private ^Tika tika (Tika.))
```
Comment (piranha) [inline src/metabase/upload.clj:223]: is this fast?

File: src/metabase/upload.clj [hunk 2]
```diff
@@ -229,25 +243,27 @@
...
-         ;; - The precedence order in how we define them, e.g.. bias towards comma
-         (sort-by (fn [[_ [header-column-count data-column-count]]]
-                    [(when header-column-count
-                       (> header-column-count 1))
-                     (= header-column-count data-column-count)
-                     header-column-count])
-                  u/reverse-compare)
-         ffirst
-         assert-inferred-separator)))
+    (if (= "tsv" extension)
```
Comment (crisptrutski) [inline src/metabase/upload.clj:246]: Off topic, but felt silly not to do it at this at this point.

File: test/metabase/upload_test.clj [hunk 3]
```diff
@@ -1115,6 +1118,47 @@
                                              :upload-seconds    pos?}}}
                    (last-audit-event :upload-create)))))))))
 
+(defn- write-1-pixel-jpg
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:1121]: I was too lazy to add a test resource file, which felt like overkill. I could be coaxed.

Comment (piranha) [reply test/metabase/upload_test.clj:1121]: Is there nothing smaller that would pose as a binary file? 🤣 I'd say if it's more than 20-30 bytes it should go in resources. :)

File: src/metabase/upload.clj (hunk 1 above)
Comment (crisptrutski) [reply src/metabase/upload.clj:223]: Haven't benchmarked it, but I'd be shocked if it's slower than the actual update. Are you worried about DoS?

Comment (crisptrutski) [reply src/metabase/upload.clj:223]: I guess we could swap the order with the auth check.

File: test/metabase/upload_test.clj (hunk 3 above)
Comment (crisptrutski) [reply test/metabase/upload_test.clj:1121]: Hmm a 1-pixel GIF is 35b... I need to rise to your challenge 😆

File: src/metabase/upload.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/upload.clj:223]: nah, I'm worried about start up time being delayed a little bit more :) As for content check - I know it's not very fast, but I don't have any good ideas. Auth check should be done before content check for sure though.

File: test/metabase/upload_test.clj (hunk 3 above)
Comment (piranha) [reply test/metabase/upload_test.clj:1121]: https://news.ycombinator.com/item?id=38878480

Comment (crisptrutski) [reply test/metabase/upload_test.clj]: Didn't need to be an image 😈

File: src/metabase/upload.clj [hunk 4]
```diff
@@ -215,12 +216,25 @@
       (throw (ex-info (tru "Unable to recognise file separator")
                       {:status-code 422}))))
 
+(def ^:private allowed-extensions #{nil "csv" "tsv" "txt"})
```
Comment (crisptrutski) [inline src/metabase/upload.clj:219]: The frontend does not allow txt or no extension, but I'm worried that there are customers using the API directly, and these are still very reasonable cases I've encountered before.

File: src/metabase/upload.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/upload.clj:223]: It's 1.5 ms, I think that's fine.

File: test/metabase/upload_test.clj [hunk 5]
```diff
@@ -1115,6 +1118,24 @@
                                              :upload-seconds    pos?}}}
                    (last-audit-event :upload-create)))))))))
 
+(defn- write-empty-gzip
+  "Writes the data for an empty gzip file"
+  [^java.io.File file]
+  (let [jpg-bytes (byte-array
```
Comment (piranha) [inline test/metabase/upload_test.clj:1124]: are they jpg bytes though? 😁

**Review by calherries — APPROVED** [review]
Nice

**Comment by crisptrutski** [comment]
Damn, auto merge fail after rebasing off another branch 🤦

**Pull request merged**